### PR TITLE
[EncoderDecoder] Make tests more aggressive

### PR DESCRIPTION
**Pull request by patrickvonplaten**
# What does this PR do?

Before merging #9183, we should make sure that the EncoderDecoder and caching tests are aggressive enough to be sure everything works as expected.

In addition, this PR refactors the `_expand_mask` function in Bart making it cleaner and move the responsibility correctly to the `attention_mask` creation.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -941,11 +929,21 @@ def forward(
             attention_mask = input_ids.ne(self.config.pad_token_id).to(torch.long)
             # never mask leading token, even if it is pad
             attention_mask[:, 0] = attention_mask[:, 1]
+            if past_key_values_length > 0:
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_bart.py:932]: `past_key_values_length` is actually not the responsibility of the `expand_mask` function but of the attention mask -> cleaner to do it here.

File: src/transformers/models/bart/modeling_bart.py [hunk 2]
```diff
@@ -106,16 +104,6 @@ def _expand_mask(
 
     expanded_mask = mask[:, None, None, :].expand(bsz, 1, tgt_len, src_len).to(dtype)
 
-    if past_key_values_length > 0:
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_bart.py:109]: better to make `_expand_mask` independent from `past_key_values_length`

File: tests/test_modeling_encoder_decoder.py [hunk 3]
```diff
@@ -568,15 +568,27 @@ def test_bert2bert_summarization(self):
...
+            pad_to_max_length=True,
+            max_length=512,
+            return_tensors="pt",
+        )
+        output_ids = model.generate(
+            input_dict["input_ids"].to(torch_device), attention_mask=input_dict["attention_mask"].to(torch_device)
+        )
+        summary = tokenizer.batch_decode(output_ids, skip_special_tokens=True)
+
+        self.assertEqual(summary, [EXPECTED_SUMMARY_SIGMA, EXPECTED_SUMMARY_AMERICA])
```
Comment (patrickvonplaten) [inline tests/test_modeling_encoder_decoder.py:591]: We need more aggressive slow tests for EncoderDecoderModel now that it's gonna be used in combination with `use_cache`
